core/src/main/java/io/undertow/util/PathTemplateRouter.java

```diff
@@ -0,0 +1,2552 @@
+                        + "templates. Use defaultTarget instead");
+
+            Builder b = this;
+            b.defaultTargetFactory = () -> defaultTarget;
+            return b;
+        }
+
+        /**
+         * Gets the existing target factory for the specified template, provided that one has already been added.
+         *
+         * @param template The template.
+         *
+         * @return The target factory or {@code NULL}.
+         */
+        public S getTemplateTarget(final Template<S> template) {
+            final PatternEqualsAdapter<Template<S>> key = new PatternEqualsAdapter<>(template);
+            return templates.get(key);
+        }
+
+        /**
+         * @return A mutable map of {@link PatternEqualsAdapter}s for all templates added to this builder.
+         */
+        public Map<PatternEqualsAdapter<Template<S>>, S> getTemplates() {
+            return templates;
+        }
+
+        /**
+         * Clears all existing templates from this builder.
+         *
+         * @return Reference to the builder.
+         */
+        public Builder<S, T> clearTemplates() {
+            templates.clear();
+            return this;
+        }
+
+        /**
+         * Adds the specified template to this builder, provided that it has not been added yet. Throws
+         * {@link IllegalArgumentException} if this builder already contains the template.
+         *
+         * @param template The template.
+         *
+         * @return Reference to the builder.
+         */
+        public Builder<S, T> addTemplate(final Template<S> template) {
+            final PatternEqualsAdapter<Template<S>> item = new PatternEqualsAdapter<>(template);
+            if (templates.containsKey(item))
+                throw new IllegalArgumentException("The builder already contains a template with the same "
+                        + "pattern for '" + template.pathTemplate + "'");
+            templates.put(item, template.target);
+            return this;
+        }
+
+        /**
+         * Adds the specified template to this builder, provided that it has not been added yet. Throws
+         * {@link IllegalArgumentException} if this builder already contains the template.
+         *
+         * @param pathTemplate  The path template.
+         * @param targetFactory The target factory.
+         *
+         * @return Reference to the builder.
+         */
+        public Builder<S, T> addTemplate(final String pathTemplate, final S targetFactory) {
+            final Template<S> template = parseTemplate(pathTemplate, targetFactory);
+            return Builder.this.addTemplate(template);
+        }
+
+        /**
+         * Removes the specified path template from this builder.
+         *
+         * @param pathTemplate The path template.
+         *
+         * @return Reference to the builder.
+         */
+        public Builder<S, T> removeTemplate(final String pathTemplate) {
```

Comment:
I can add a map for the path template string -> adapter instance, then look up the key and remove it directly from the "templates" map by reference. However, there are two things to consider:
1. This is part of the "setup phase" of the routers that typically happens only once when a service starts up.  So the overhead of parsing again has a very small impact on the overall performance. I have personally never seen someone add a path template and then remove it again. I have added this method to support the "remove" method that already existed in PathTemplateHandler - so basically for backwards compatibility. In all of our projects we create new routers using new builders if we do want to mutate the routes, so I suspect that this will almost never be used and for that reason the overhead is probably okay if you consider the next point.
2. Consider the following two template strings /books/{bookId}/chapters and /books/{id}/chapters These two strings are not equal, but the patterns represented by the templates are equal. Since one can't have "duplicate" patterns in a router, I would argue that the intention of removing the template is to remove the pattern associated with it so that you can potentially replace it with another template that uses a different parameter name - for example. So if we were to implement that map to remove by reference and we can't find the exact string in the map, then we would have to parse the string again anyway in order to be sure that it has been removed.